### PR TITLE
Task Dependency

### LINKED ISSUE #80: Add Dependency Command
Command to specify dependency between two tasks, where the dependee needs to be completed before the dependant.

### LINKED ISSUE #81: Specify Dependency Field
Tasks contain notion of dependency

### PR DESCRIPTION
**Pull request by pengnam**
## PR introduces notion of dependency into tasks. 
Resolves #80 Resolves #81 
## Two main segments:
1) Dependency field in Task model
* dependency is identified using String hashes of task. Refer to Dependency class  [https://github.com/CS2103-AY1819S1-F11-3/main/blob/master/src/main/java/seedu/address/model/task/Dependency.java](url) to check details of usage.
* ALTERNATIVE: dependency specified using Task objects will introduce unnecessary complication.Tasks are immutable and need to be swapped out and replaced when dependencies are updated. Storage of tasks is also challenging  

2) Dependency command 
Dependencies can be specified using Task index in list. (Similar to edit and delete command). 
Refer to [https://github.com/CS2103-AY1819S1-F11-3/main/blob/master/src/main/java/seedu/address/logic/commands/DependencyCommand.java](url) for more details.

## Additional notes in PR:
1) hashcode is now **unique** and **consistent**
- hashcode now uniquely and consistently identifies a task(Refer to hashcode function in Task class).
- was previously inconsistent after previous PR changes when introducing new fields
2) TaskCard now displays hash of task and hashes of tasks that the task is dependent on. 

## Files Edited:
Other files edited in process of creating this PR. A large number of these files are edited after Task function signature change.
src/main/java/seedu/address/commons/events/ui/TaskPanelSelectionChangedEvent.java
src/main/java/seedu/address/logic/Logic.java
src/main/java/seedu/address/logic/LogicManager.java
src/main/java/seedu/address/logic/commands/AddCommand.java
src/main/java/seedu/address/logic/commands/CompleteCommand.java
src/main/java/seedu/address/logic/commands/DeleteCommand.java
src/main/java/seedu/address/logic/commands/DependencyCommand.java
src/main/java/seedu/address/logic/commands/EditCommand.java
src/main/java/seedu/address/logic/commands/FindCommand.java
src/main/java/seedu/address/logic/commands/SelectCommand.java
src/main/java/seedu/address/logic/parser/AddCommandParser.java
src/main/java/seedu/address/logic/parser/AddressBookParser.java
src/main/java/seedu/address/logic/parser/CompleteCommandParser.java
src/main/java/seedu/address/logic/parser/DependencyCommandParser.java
src/main/java/seedu/address/logic/parser/FindCommandParser.java
src/main/java/seedu/address/logic/parser/ParserUtil.java
src/main/java/seedu/address/model/Model.java
src/main/java/seedu/address/model/ModelManager.java
src/main/java/seedu/address/model/ReadOnlyTaskManager.java
src/main/java/seedu/address/model/TaskManager.java
src/main/java/seedu/address/model/task/Dependency.java
src/main/java/seedu/address/model/task/Description.java
src/main/java/seedu/address/model/task/DueDate.java
src/main/java/seedu/address/model/task/Name.java
src/main/java/seedu/address/model/task/NameContainsKeywordsPredicate.java
src/main/java/seedu/address/model/task/PriorityValue.java
src/main/java/seedu/address/model/task/Status.java
src/main/java/seedu/address/model/task/Task.java
src/main/java/seedu/address/model/task/UniqueTaskList.java
src/main/java/seedu/address/model/task/exceptions/DuplicateTaskException.java
src/main/java/seedu/address/model/task/exceptions/TaskNotFoundException.java
src/main/java/seedu/address/model/util/SampleDataUtil.java
src/main/java/seedu/address/storage/XmlAdaptedLabel.java
src/main/java/seedu/address/storage/XmlAdaptedTask.java
src/main/java/seedu/address/storage/XmlSerializableTaskManager.java
src/main/java/seedu/address/ui/BrowserPanel.java
src/main/java/seedu/address/ui/TaskCard.java
src/main/java/seedu/address/ui/TaskListPanel.java
src/main/resources/view/PersonListCard.fxml
src/test/data/XmlSerializableTaskManagerTest/typicalTasksTaskManager.xml
src/test/java/guitests/guihandles/TaskCardHandle.java
src/test/java/guitests/guihandles/TaskListPanelHandle.java
src/test/java/seedu/address/commons/util/XmlUtilTest.java
src/test/java/seedu/address/logic/commands/AddCommandIntegrationTest.java
src/test/java/seedu/address/logic/commands/AddCommandTest.java
src/test/java/seedu/address/logic/commands/CommandTestUtil.java
src/test/java/seedu/address/logic/commands/CompleteCommandTest.java
src/test/java/seedu/address/logic/commands/DeleteCommandTest.java
src/test/java/seedu/address/logic/commands/DependencyCommandTest.java
src/test/java/seedu/address/logic/commands/EditCommandTest.java
src/test/java/seedu/address/logic/commands/FindCommandTest.java
src/test/java/seedu/address/logic/parser/AddCommandParserTest.java
src/test/java/seedu/address/logic/parser/AddressBookParserTest.java
src/test/java/seedu/address/logic/parser/CompleteCommandParserTest.java
src/test/java/seedu/address/logic/parser/EditCommandParserTest.java
src/test/java/seedu/address/logic/parser/FindCommandParserTest.java
src/test/java/seedu/address/logic/parser/ParserUtilTest.java
src/test/java/seedu/address/model/ModelManagerTest.java
src/test/java/seedu/address/model/TaskManagerTest.java
src/test/java/seedu/address/model/task/DependencyTest.java
src/test/java/seedu/address/model/task/DescriptionTest.java
src/test/java/seedu/address/model/task/DueDateTest.java
src/test/java/seedu/address/model/task/NameContainsKeywordsPredicateTest.java
src/test/java/seedu/address/model/task/NameTest.java
src/test/java/seedu/address/model/task/PriorityValueTest.java
src/test/java/seedu/address/model/task/StatusTest.java
src/test/java/seedu/address/model/task/TaskTest.java
src/test/java/seedu/address/model/task/UniqueTaskListTest.java
src/test/java/seedu/address/storage/XmlAdaptedTaskTest.java
src/test/java/seedu/address/testutil/EditTaskDescriptorBuilder.java
src/test/java/seedu/address/testutil/TaskBuilder.java
src/test/java/seedu/address/testutil/TaskManagerBuilder.java
src/test/java/seedu/address/testutil/TaskUtil.java
src/test/java/seedu/address/testutil/TestUtil.java
src/test/java/seedu/address/testutil/TypicalIndexes.java
src/test/java/seedu/address/testutil/TypicalTasks.java
src/test/java/seedu/address/ui/TaskCardTest.java
src/test/java/seedu/address/ui/TaskListPanelTest.java
src/test/java/seedu/address/ui/testutil/GuiTestAssert.java
src/test/java/systemtests/AddCommandSystemTest.java
src/test/java/systemtests/AddressBookSystemTest.java
src/test/java/systemtests/ClearCommandSystemTest.java
src/test/java/systemtests/DeleteCommandSystemTest.java
src/test/java/systemtests/EditCommandSystemTest.java
src/test/java/systemtests/FindCommandSystemTest.java
src/test/java/systemtests/ModelHelper.java
src/test/java/systemtests/SampleDataTest.java
src/test/java/systemtests/SelectCommandSystemTest.java

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/DependencyCommand.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+                dependeeTask.getDescription(),
+                dependeeTask.getLabels(),
+                dependeeTask.getStatus(),
+                dependeeTask.getDependency().addDependency(dependantTask)
+        );
+    }
+
+
+
+
```
Comment (unsatisfiedpopcorn) [inline src/main/java/seedu/address/logic/commands/DependencyCommand.java]: Should not have multiple blank lines

File: src/main/java/seedu/address/model/task/Dependency.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+    private Set<String> value = new HashSet<>();
+
+    /**
+     * Constructs an {@code Dependency}.
+     *
+     * @param hashes A list of hashes of task dependencies.
+     */
+
+    public Dependency(Set<String> hashes) {
```
Comment (unsatisfiedpopcorn) [inline src/main/java/seedu/address/model/task/Dependency.java]: Should not have blank line separating function and javadocs

File: src/main/java/seedu/address/model/task/Task.java [hunk 3]
```diff
@@ -24,11 +24,13 @@
     // Data fields
     private final Description description;
     private final Set<Label> labels = new HashSet<>();
+    private final Dependency dependency;
 
     /**
      * Every field must be present and not null.
      * Status of a new task is initialized to in progress unless specified.
      */
+    //TODO: Remove the need for multiple constructors
```
Comment (unsatisfiedpopcorn) [inline src/main/java/seedu/address/model/task/Task.java]: Unfinished TODO

File: src/main/java/seedu/address/model/task/Task.java [hunk 4]
```diff
@@ -78,6 +82,11 @@ public Status getStatus() {
         return status;
     }
 
+    public Dependency getDependency() {
+        return dependency;
+    }
+
+
     public boolean isCompleted() {
```
Comment (unsatisfiedpopcorn) [inline src/main/java/seedu/address/model/task/Task.java]: Single blank line would be preferred

File: src/main/java/seedu/address/storage/XmlAdaptedTask.java [hunk 5]
```diff
@@ -97,6 +106,7 @@ public XmlAdaptedTask(Task source) {
      * @throws IllegalValueException if there were any data constraints violated in the adapted task
      */
     public Task toModelType() throws IllegalValueException {
+        //TODO: add dependency parser
```
Comment (unsatisfiedpopcorn) [inline src/main/java/seedu/address/storage/XmlAdaptedTask.java]: TODO:

File: src/main/java/seedu/address/ui/TaskCard.java [hunk 6]
```diff
@@ -39,7 +39,12 @@
     @FXML
     private Label status;
     @FXML
+    private Label hash;
+    @FXML
+    private Label dependency;
+    @FXML
     private FlowPane tags;
+    //TODO:Cannot find local vairable tag
```
Comment (unsatisfiedpopcorn) [inline src/main/java/seedu/address/ui/TaskCard.java]: TODO:

File: src/test/java/seedu/address/logic/commands/CommandTestUtil.java [hunk 7]
```diff
@@ -166,6 +166,27 @@ public static void showTaskAtIndex(Model model, Index targetIndex) {
 
         assertEquals(1, model.getFilteredTaskList().size());
     }
+    /**
+     * Updates {@code model}'s filtered list to show only the two tasks at {@code firstTargetIndex}
+     * and {@code secondTargetIndex}
+     * in the
+     * {@code model}'s task manager.
```
Comment (unsatisfiedpopcorn) [inline src/test/java/seedu/address/logic/commands/CommandTestUtil.java]: Can afford to remove the line breaks for readability

File: src/test/java/seedu/address/logic/commands/DependencyCommandTest.java [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+import seedu.address.commons.core.index.Index;
+import seedu.address.logic.CommandHistory;
+import seedu.address.model.Model;
+import seedu.address.model.ModelManager;
+import seedu.address.model.UserPrefs;
+import seedu.address.model.task.Task;
+
+
+
+public class DependencyCommandTest {
```
Comment (unsatisfiedpopcorn) [inline src/test/java/seedu/address/logic/commands/DependencyCommandTest.java]: Is there a test for equality?

File: src/test/java/seedu/address/model/task/DependencyTest.java [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+import java.util.HashSet;
+import java.util.Set;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import seedu.address.testutil.TaskBuilder;
+
+
+public class DependencyTest {
```
Comment (unsatisfiedpopcorn) [inline src/test/java/seedu/address/model/task/DependencyTest.java]: Should test equality if possible

File: src/test/java/seedu/address/logic/commands/DependencyCommandTest.java (hunk 8 above)
Comment (pengnam) [reply src/test/java/seedu/address/logic/commands/DependencyCommandTest.java]: Just added

File: src/test/java/seedu/address/model/task/DependencyTest.java (hunk 9 above)
Comment (pengnam) [reply src/test/java/seedu/address/model/task/DependencyTest.java]: Just added

**Pull request merged**